### PR TITLE
8241403: JavaThread::get_thread_name() should be ThreadSMR-aware

### PR DESCRIPTION
**Pull request by dcubed-ojdk**
A minor fix to add a new function:

    bool Thread::is_JavaThread_protected(const JavaThread* p)

that returns true when the target JavaThread* is protected and false
otherwise. Update JavaThread::get_thread_name() to create a
ThreadsListHandle and use the new is_JavaThread_protected(). Also
update JvmtiTrace::safe_get_thread_name() to use the new
is_JavaThread_protected().

This fix is tested via a Mach5 Tier[1-8] run.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8241403](https://bugs.openjdk.java.net/browse/JDK-8241403): JavaThread::get_thread_name() should be ThreadSMR-aware

### Reviewers
 * [Robbin Ehn](https://openjdk.java.net/census#rehn) (@robehn - **Reviewer**) ⚠️ Review applies to 4298bb4dcf7a0eeab67da32e541149079be74ceb
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 00e3886810d2509d6f2a89ad887cdc3a1810fd22
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 4298bb4dcf7a0eeab67da32e541149079be74ceb

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2535/head:pull/2535`
`$ git checkout pull/2535`

### CODE REVIEW

**Comment by dcubed-ojdk** [comment]
@dholmes-ora and @fisk - It looks like both of you are interested
in this fix. @robehn - This one is Thread-SMR related so you might
also want to take a look.

**Comment by dcubed-ojdk** [comment]
@sspitsyn - I'm making a minor JVM/TI tracing tweak here so this
might also interest you.

**Review by robehn — APPROVED** [review]
Seems alright.

**Comment by dcubed-ojdk** [comment]
@robehn - Thanks for the review!

File: src/hotspot/share/runtime/thread.hpp [hunk 1]
```diff
@@ -1687,6 +1689,7 @@ class JavaThread: public Thread {
   void verify();
   const char* get_thread_name() const;
  protected:
+  friend class JvmtiTrace;  // so get_thread_name_string() can be called
```
Comment (coleenp) [inline src/hotspot/share/runtime/thread.hpp:1692]: I was trying to think of a way to not have JvmtiTrace not be a friend of JavaThread for this, maybe by adding a default value parameter to return "\<NOT FILLED IN\>" rather than Thread::name.
is_JavaThread_protected only seems to be called by JvmtiTrace also, so should be private (with the friend, which also makes the friend unfortunate).

**Review by coleenp — CHANGES_REQUESTED** [review]
This looks good, I have a change - could you see if you agree.
It's nice not to have to take Threads_lock to get the name of the thread.

File: src/hotspot/share/runtime/thread.hpp (hunk 1 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.hpp:1692]: JavaThread::get_thread_name() also calls is_JavaThread_protected().

The "friend" is so that JvmtiTrace can call get_thread_name_string()
and we can get rid of the JvmtiTrace version of the logic. I kept the
"<NOT FILLED IN>" rather than figure out a way to call Thread::name()
so we don't introduce the possibility of a compatibility issue for any
code that might depend on that hand rolled string value...

Comment (coleenp) [reply src/hotspot/share/runtime/thread.hpp:1692]: So Thread::is_JavaThread_protected() should be "protected" then, not public.
yes, I was suggesting adding a default last parameter like
    JavaThread::get_thread_name(char* default = Thread::name());
and pass "\<NOT_FILLED_IN\>" from JVMTI. Then JVMTI doesn't have to be a friend and have more visibility to the JavaThread class than it should have.

**Review by sspitsyn — APPROVED** [review]
Hi Dan,
It looks good to me modulo your your discussion with Coleen on JvmtiTrace being a friend.
Thanks,
Serguei

File: src/hotspot/share/runtime/thread.hpp (hunk 1 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.hpp:1692]: Ahhh.... I think I finally understand what you mean...
I'll look in the AM after I've had some coffee... :-)

Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.hpp:1692]: The name "default" is reserved so I went with "def_name".
You can't call Thread::name() from the declaration:
```
./open/src/hotspot/share/runtime/thread.hpp:1690:62: error: call to non-static member function without an object argument
  const char* get_thread_name(const char* def_name = Thread::name()) const;
                                                     ~~~~~~~~^~~~
```
so I went with setting `def_name = NULL` in the declaration and decoding that state in get_thread_name():
` return (def_name != NULL) ? def_name : Thread::name();`

Comment (coleenp) [reply src/hotspot/share/runtime/thread.hpp:1692]: default_name, it's not long enough to abbreviate...  but yes. thanks!

Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.hpp:1692]: And it looks like making is_JavaThread_protected() a protected function doesn't work either:

Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.hpp:1692]: `JavaThread::get_thread_name()` is calling `is_JavaThread_protected()` on the calling Thread and NOT on the `this` JavaThread so `protected` doesn't work.

**Comment by dcubed-ojdk** [comment]
Mach5 Tier[1-8] testing on CR1 is in process:
- Mach5 Tier[1-3] - no failures
- Mach5 Tier4 - 2 unrelated, known failures - still running
- Mach5 Tier5 - 1 unrelated, known failure
- Mach5 Tier6 - no failures - still running
- Mach5 Tier7 - just started
- Mach5 Tier8 - just started

@robehn, @coleenp, @sspitsyn - please take a look when you get the chance.

**Review by sspitsyn — APPROVED** [review]
The update looks good.
Thanks,
Serguei

**Review by coleenp — APPROVED** [review]
Thanks!

File: src/hotspot/share/runtime/thread.hpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/runtime/thread.hpp]: ok.

**Comment by dcubed-ojdk** [comment]
@coleenp and @sspitsyn - Thanks for the re-reviews!

**Comment by dcubed-ojdk** [comment]
@dholmes and @fisk - did either or both of you plan to review this one?

**Comment by dcubed-ojdk** [comment]
Merged with master and retested with Mach5 Tier1.

File: src/hotspot/share/prims/jvmtiTrace.cpp [hunk 2]
```diff
@@ -277,15 +277,10 @@ const char *JvmtiTrace::safe_get_thread_name(Thread *thread) {
...
-  }
-  oop name = java_lang_Thread::name(threadObj);
-  if (name == NULL) {
-    return "<NOT FILLED IN>";
-  }
-  return java_lang_String::as_utf8_string(name);
+
+  // If the target JavaThread is not protected, then we return the
+  // specified non-NULL string:
+  return thread->as_Java_thread()->get_thread_name("<NOT FILLED IN>");
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiTrace.cpp:283]: That's not what the original logic is doing. We only return "NOT FILLED IN" if we find that the java.lang.Thread oop does not have a name set. That's not normally possible but this code is intended to be 100% safe, hence the check. (And it may be possible to see null during the attach process ...)

I would expect the target thread to already be protected when this method is called in any case. So while this code partially duplicates JavaThread::get_thread_name() I'm not convinced we need to change this code to use JavaThread::get_thread_name() instead.

File: src/hotspot/share/runtime/thread.cpp [hunk 3]
```diff
@@ -482,6 +482,27 @@ void Thread::check_for_dangling_thread_pointer(Thread *thread) {
 }
 #endif
 
+// Is the target JavaThread protected by this Thread:
+bool Thread::is_JavaThread_protected(const JavaThread* p) {
+  if (current() == p) {
+    // Current thread is always protected:
+    return true;
+  }
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.cpp:490]: This is true but it is awkward that we are going to manifest Thread::current multiple times when using this - see below.

File: src/hotspot/share/runtime/thread.cpp [hunk 4]
```diff
@@ -2519,21 +2540,20 @@ void JavaThread::verify() {
...
-#ifdef ASSERT
-  // early safepoints can hit while current thread does not yet have TLS
-  if (!SafepointSynchronize::is_at_safepoint()) {
-    // Current JavaThreads are allowed to get their own name without
-    // the Threads_lock.
-    if (Thread::current() != this) {
-      assert_locked_or_safepoint_or_handshake(Threads_lock, this);
-    }
+const char* JavaThread::get_thread_name(const char* default_name) const {
+  Thread* thread = Thread::current();
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.cpp:2547]: So you manifest Thread::current() and then is_JavaThread_protected will manifest it again.

File: src/hotspot/share/runtime/thread.cpp [hunk 5]
```diff
@@ -2519,21 +2540,20 @@ void JavaThread::verify() {
...
+  ThreadsListHandle tlh(thread);
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.cpp:2548]: I would only expect this code to create a TLH if the target is not protected, otherwise why do we need to search through the existing thread-lists as it would suffice to see if the target is within the newly created TLH ??
I expected the basic logic here to be:
`if (thread is protected) return name;
else { protect the thread; return name; }`

though if it is cheaper to create a TLH than to search the existing ones, then we may as well just create the TLH, check that the target has been captured and then return its name.

That said we need to carefully examine where the taking of the Threads_lock would occur in the thread termination sequence, to understand what observable states are possible with respect to the target having a threadObj. (Though this may have already changed with the introduction of Thread-SMR.)

That all said I'm not so sure the only thing we actually need to do for this issue is to replace the use of Threads_lock, in the caller, with suitable TLH usage, and then in get_thread_name() assert that the target is in fact protected.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Dan,

Sorry but I have a lot of issues with this. After thinking about it a lot I don't think the current approach is what is needed. To repeat what I wrote in one of the comments I think the simple fix here is to replace the use of Threads_lock in the caller with suitable use of a TLH, and then replace the assert_locked_or_safepoint(Threads_lock) with assert(is_JavaThread_protected(t)) where is_JavaThread_protected checks the target for either being the current thread or included in a TLH of the current thread. I don't think get_thread_name() should try to protect the target as that is the responsibility of the caller.

Thanks,
David

**Comment by dcubed-ojdk** [comment]
@dholmes-ora - I'm really glad that I waited for your review! Thanks for taking the time.

File: src/hotspot/share/runtime/thread.cpp [hunk 6]
```diff
@@ -482,6 +482,27 @@ void Thread::check_for_dangling_thread_pointer(Thread *thread) {
 }
 #endif
 
+// Is the target JavaThread protected by this Thread:
+bool Thread::is_JavaThread_protected(const JavaThread* p) {
```
[the remark this answers is not in the document]
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.cpp:486]: I made it a function on Thread because a Thread can use a ThreadsListHandle
to protect access to JavaThreads and we're checking the ThreadsList(s) that
are associated with the calling Thread.

It dawns on me that the comment could have been more clear:

`// Is the target JavaThread protected by the calling Thread:`

If I had written the comment that way, then I would have realized that the
function should be static in Thread and then it could call Thread::current()
itself and then make its checks of the ThreadsList(s).

File: src/hotspot/share/runtime/thread.cpp (hunk 5 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.cpp:2548]: I dropped the TLH into JavaThread::get_thread_name() because I was focused
on protecting the fetching of the name from the JavaThread. In reality that
doesn't happen in a vacuum so a TLH should really exist in the caller or some
where higher in the call stack.

There is a subtle point here: Creating a TLH in JavaThread::get_thread_name()
may not actually protect "this" JavaThread. "this" JavaThread would only be
protected by a newly created TLH, if it is on the current ThreadsList. If we
happen to call java_thread->get_thread_name() too early in the lifecycle or
too late in the lifecycle, then that newly created TLH doesn't help.

The original code's assertion check was a little complicated, but still was
a good idea:

    if we're not at a safepoint and if we're not calling the function on our self
       then assert_locked_or_safepoint_or_handshake(Threads_lock)

I'm going to add back a simpler assertion that takes advantage of the
fact that Thread::is_JavaThread_protected() returns true when the
target is the current thread.

File: src/hotspot/share/prims/jvmtiTrace.cpp (hunk 2 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/prims/jvmtiTrace.cpp:283]: My intent was to have safe_get_thread_name() take advantage of
JavaThread::get_thread_name() when we know that we're dealing
with a JavaThread. Shared code means less maintenance, but it
does look like I didn't achieve equivalence so I'm going to back
out my changes for JvmtiTrace::safe_get_thread_name().

File: src/hotspot/share/compiler/compileBroker.cpp [hunk 7]
```diff
@@ -1005,7 +1005,6 @@ void CompileBroker::init_compiler_sweeper_threads() {
       _compilers[1]->set_num_compiler_threads(i + 1);
       if (TraceCompilerThreads) {
         ResourceMark rm;
-        MutexLocker mu(Threads_lock);
```
[the remark this answers is not in the document]
Comment (dcubed-ojdk) [reply src/hotspot/share/compiler/compileBroker.cpp:1008]: Ouch! You are correct that any time after the make_thread() call,
the newly created JavaThread could run and exit quickly. In the
baseline code, the Threads_lock grab only keeps 'ct' from exiting
while the lock is held. It doesn't help if 'ct' exited before the lock
was grabbed.

I'm not sure why you would think that 'ct' could be protected by
the current call chain. Of course, I could be missing something.

I'm going to see if I can do something here with a TLH.

**Comment by dcubed-ojdk** [comment]
The changes in CR2 are:

- Make Thread::is_JavaThread_protected() static and make it clear that
  it checks to see if the calling thread is protecting the target;
  added VMThread access as always protected.
- Revert changes in JvmtiTrace::safe_get_thread_name().
- Revert the JavaThread::get_thread_name() default_name parameter and
  don't create a ThreadsListHandle in the function; restore a simpler
  assert for the case where the target thread is not protected.
- Add Threads::threads_do() and Threads::java_threads_do() that work
  with a JavaThreadIteratorWithHandle.
- Update src/hotspot/share/services/management.cpp to switch from using
  the Threads_lock to using JavaThreadIteratorWithHandle.
- Redo compileBroker.cpp changes to use a ThreadsListHandle to protect 
  the CompilerThread/JavaThread.
- Remove a stale comment from src/hotspot/share/prims/jvm.cpp.

You'll notice some temporary testing code in JavaThread::get_thread_name():

```
  guarantee(false, "current_thread=" INTPTR_FORMAT
            " is not protecting this=" INTPTR_FORMAT,
            p2i(current_thread), p2i(this));
```

I'm doing my current Mach5 Tier[1-8] run with this guarantee() in place.
It hasn't fired in Mach5 Tier[1-7]; Tier8 is still running, but it hasn't
fired yet. What this means is that we don't currently have a test case that
always exercises the unprotected path; we might have a test case that 
sometimes exercises the unprotected path depending on timing and/or
options, but nothing so far. My intention with this fix is not to leave any
unprotected paths for JavaThread::get_thread_name() and these test results
give me confidence, but not complete assurance.

**Comment by dcubed-ojdk** [comment]
Here's another thing that I noticed and want to discuss with the reviwers:

I added a new comment above the simpler assert that I restored into
JavaThread::get_thread_name():

```
  // Note: Since this JavaThread isn't protected by a TLH, the call to
  // this->is_handshake_safe_for() may crash, but we have debug bits so...
  assert(SafepointSynchronize::is_at_safepoint() ||
         this->is_handshake_safe_for(current_thread), "JavaThread="
         INTPTR_FORMAT " is not protected, not at a safepoint and "
         "not handshake safe.", p2i(this));
```

The baseline code used:

  `assert_locked_or_safepoint_or_handshake(Threads_lock, this);`

and that function looks like this:

```
void assert_locked_or_safepoint_or_handshake(const Mutex* lock, const JavaThread* thread) {
  if (thread->is_handshake_safe_for(Thread::current())) return;
  assert_locked_or_safepoint(lock);
}
```

and is_handshake_safe_for() looks like this:

```
  // A JavaThread can always safely operate on it self and other threads
  // can do it safely if they are the active handshaker.
  bool is_handshake_safe_for(Thread* th) const {
    return _handshake.active_handshaker() == th || this == th;
  }
```

so the baseline code is also using "this" JavaThread without knowing that
it is safe. The interesting thing about is_handshake_safe_for() is that it
*is safe* if the target is the active handshaker for the current thread.
However, if the target is not the active handshaker for the current thread,
then calling `thread->is_handshake_safe_for()` and accessing the
`_handshake` field in that JavaThread may not be safe.

**Comment by dcubed-ojdk** [comment]
The CR2 changes are being tested with Mach5 Tier[1-8]:


    Mach5 Tier[1245] - no failures
    Mach5 Tier6 - 2 unrelated, known failures
    Mach5 Tier7 - 3 unrelated, known failures
    Mach5 Tier8 - still running, no failures so far

File: src/hotspot/share/compiler/compileBroker.cpp [hunk 8]
```diff
@@ -1002,6 +1003,8 @@ void CompileBroker::init_compiler_sweeper_threads() {
     if (!UseDynamicNumberOfCompilerThreads || i == 0) {
       JavaThread *ct = make_thread(compiler_t, thread_handle, _c2_compile_queue, _compilers[1], THREAD);
       assert(ct != NULL, "should have been handled for initial thread");
+      ThreadsListHandle tlh;
+      assert(tlh.includes(ct), "ct=" INTPTR_FORMAT " exited unexpectedly.", p2i(ct));
```
Comment (dholmes-ora) [inline src/hotspot/share/compiler/compileBroker.cpp:1007]: This case seems to be one where we are creating an immortal compiler thread (either because they are all immortal, or this is the first compiler thread and it is always immortal (verify that?)) - so the assert seems fine.

File: src/hotspot/share/compiler/compileBroker.cpp [hunk 9]
```diff
@@ -1021,6 +1024,8 @@ void CompileBroker::init_compiler_sweeper_threads() {
     if (!UseDynamicNumberOfCompilerThreads || i == 0) {
       JavaThread *ct = make_thread(compiler_t, thread_handle, _c1_compile_queue, _compilers[0], THREAD);
       assert(ct != NULL, "should have been handled for initial thread");
+      ThreadsListHandle tlh;
+      assert(tlh.includes(ct), "ct=" INTPTR_FORMAT " exited unexpectedly.", p2i(ct));
```
Comment (dholmes-ora) [inline src/hotspot/share/compiler/compileBroker.cpp:1028]: Ditto

File: src/hotspot/share/compiler/compileBroker.cpp [hunk 10]
```diff
@@ -1106,6 +1111,8 @@ void CompileBroker::possibly_add_compiler_threads(Thread* THREAD) {
 #endif
       JavaThread *ct = make_thread(compiler_t, compiler2_object(i), _c2_compile_queue, _compilers[1], THREAD);
       if (ct == NULL) break;
+      ThreadsListHandle tlh;
+      assert(tlh.includes(ct), "ct=" INTPTR_FORMAT " exited unexpectedly.", p2i(ct));
```
Comment (dholmes-ora) [inline src/hotspot/share/compiler/compileBroker.cpp:1115]: This seems to be the fully dynamic case. Here the TLH is potentially racing with thread termination and so may or may not capture ct. Rather than an assertion the call to  `get_thread_name()` below should be guarded by `tlh.includes(ct)`.

File: src/hotspot/share/compiler/compileBroker.cpp [hunk 11]
```diff
@@ -1125,6 +1132,8 @@ void CompileBroker::possibly_add_compiler_threads(Thread* THREAD) {
     for (int i = old_c1_count; i < new_c1_count; i++) {
       JavaThread *ct = make_thread(compiler_t, compiler1_object(i), _c1_compile_queue, _compilers[0], THREAD);
       if (ct == NULL) break;
+      ThreadsListHandle tlh;
+      assert(tlh.includes(ct), "ct=" INTPTR_FORMAT " exited unexpectedly.", p2i(ct));
```
Comment (dholmes-ora) [inline src/hotspot/share/compiler/compileBroker.cpp:1136]: Ditto - the assert is not appropriate for this case.

File: src/hotspot/share/runtime/thread.cpp [hunk 12]
```diff
@@ -482,16 +482,17 @@ void Thread::check_for_dangling_thread_pointer(Thread *thread) {
 }
 #endif
 
-// Is the target JavaThread protected by this Thread:
+// Is the target JavaThread protected by the calling Thread:
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.cpp:485]: s/calling/current/ ?

We could also expand this to clarify that even if a target JavaThread is not explicitly protected by the current thread it can still be safe to access if the two threads engage in a suitable protocol.

File: src/hotspot/share/runtime/thread.cpp [hunk 13]
```diff
@@ -2543,17 +2544,33 @@ void JavaThread::verify() {
...
+  //           " is not protecting this=" INTPTR_FORMAT,
+  //           p2i(current_thread), p2i(this));
+  guarantee(false, "current_thread=" INTPTR_FORMAT
+            " is not protecting this=" INTPTR_FORMAT,
+            p2i(current_thread), p2i(this));
+
+  // Note: Since this JavaThread isn't protected by a TLH, the call to
+  // this->is_handshake_safe_for() may crash, but we have debug bits so...
+  assert(SafepointSynchronize::is_at_safepoint() ||
+         this->is_handshake_safe_for(current_thread), "JavaThread="
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.cpp:2568]: I agree this is a pre-existing bug as accessing `this` may not be safe if the target is not protected. But if we crash attempting the access that implies the assert failed so ...

File: src/hotspot/share/services/management.cpp [hunk 14]
```diff
@@ -844,8 +844,8 @@ void VmThreadCountClosure::do_thread(Thread* thread) {
 static jint get_vm_thread_count() {
   VmThreadCountClosure vmtcc;
   {
-    MutexLocker ml(Threads_lock);
-    Threads::threads_do(&vmtcc);
+    JavaThreadIteratorWithHandle jtiwh;
+    Threads::threads_do(&jtiwh, &vmtcc);
```
Comment (dholmes-ora) [inline src/hotspot/share/services/management.cpp:848]: I'm not sure this is at all necessary. Threads::add and Threads::remove still use the Threads_lock so this code is safe as it stands. Switching to  use JTIWH may be an alternative, and may even be desirable, but that hasn't been established, and is not necessary for safety and so seems outside the scope of this bug IMO.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Dan,

Still a few comments/concerns - see below.

Thanks,
David

File: src/hotspot/share/runtime/thread.cpp (hunk 13 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.cpp:2568]: ... that will tell us that we have a spot in the code where a TLH is
missing once we've analyzed the crash. So are you okay with this
assert() having a bug since the original assert has the same bug?

File: src/hotspot/share/services/management.cpp (hunk 14 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/services/management.cpp:848]: Holding the Threads_lock is no longer being considered as
proper protection for the JavaThread::get_thread_name() call
so I had to change the code to use a JavaThreadIteratorWithHandle
so that we have the protection of the TLH. That closure eventually
gets us down to a JavaThread::get_thread_name() call...

**Review by dcubed-ojdk — COMMENTED** [review]
@dholmes-ora - thanks for the re-review!

File: src/hotspot/share/runtime/thread.cpp [hunk 15]
```diff
@@ -2907,6 +2950,11 @@ void Threads::threads_do(ThreadClosure* tc) {
   non_java_threads_do(tc);
 }
 
+void Threads::threads_do(JavaThreadIteratorWithHandle* jtiwh_p, ThreadClosure* tc) {
+  java_threads_do(jtiwh_p, tc);
+  non_java_threads_do(tc);
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.cpp:2955]: You should still have

```assert_locked_or_safepoint(Threads_lock);```

for non-JavaThread access.

File: src/hotspot/share/services/management.cpp [hunk 16]
```diff
@@ -1705,10 +1700,10 @@ JVM_ENTRY(jint, jmm_GetInternalThreadTimes(JNIEnv *env,
 
   ThreadTimesClosure ttc(names_ah, times_ah);
   {
-    MutexLocker ml(THREAD, Threads_lock);
-    Threads::threads_do(&ttc);
+    JavaThreadIteratorWithHandle jtiwh;
+    Threads::threads_do(&jtiwh, &ttc);
```
Comment (dholmes-ora) [inline src/hotspot/share/services/management.cpp:1704]: Same comments as above re need for Threads_lock.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Dan,

Sorry still some issue here.

Okay let me back up again. The old code for get_thread_name() thinks it is only safe to access the name of another thread if we hold the Threads_lock or we are at a safepoint. That isn't true as I've pointed out before - there are other ways for the access to be safe. And as a result of this, code calling get_thread_name() would acquire the Threads_lock just to get past the assertion e.g the CompilerBroker code. I've now ascertained that the CompilerBroker calls to get_thread_name() are perfectly safe by construction due to holding the CompilerThread_lock and/or dealing with an immortal thread.

The intent of filing this issue was for get_thread_name() to also (instead?) recognise that it is safe if the target is protected by a TLH - thus relaxing the original constraint by adding another checkable condition - and to recognise that a handshake also makes things safe. I didn't recognise at the time that these are still overly strong requirements.

So the change you are proposing is to replace the "needs the Threads_lock" condition with "must have an active TLH" condition when calling get_thread_name(). So now the unnecessary use of the Threads_lock in the CompileBroker is replaced by unnecessary use of the TLH - which is unfortunate IMO but a fix for that is out of scope for this change.

Also the management code changes now require use of a TLH  - but there is a bug in replacing the Threads_lock usage: see below!

As a Threads_lock->TLH conversion for users of get_thread_name() I can accept the change in its current form. But it needs to restore the asserts in the CompileBroker code as commented below and fixing the Threads_lock bug in the management code.

Thanks,
David

File: src/hotspot/share/compiler/compileBroker.cpp [hunk 17]
```diff
@@ -1109,10 +1111,11 @@ void CompileBroker::possibly_add_compiler_threads(Thread* THREAD) {
 #endif
       JavaThread *ct = make_thread(compiler_t, compiler2_object(i), _c2_compile_queue, _compilers[1], THREAD);
       if (ct == NULL) break;
+      ThreadsListHandle tlh;
+      if (!tlh.includes(ct)) break;
```
[the remark this answers is not in the document]
Comment (dcubed-ojdk) [reply src/hotspot/share/compiler/compileBroker.cpp:1115]: Okay. I've switched this one back to an assertion after the TLH.
Thanks for the analysis.

File: src/hotspot/share/compiler/compileBroker.cpp [hunk 18]
```diff
@@ -1129,10 +1132,11 @@ void CompileBroker::possibly_add_compiler_threads(Thread* THREAD) {
     for (int i = old_c1_count; i < new_c1_count; i++) {
       JavaThread *ct = make_thread(compiler_t, compiler1_object(i), _c1_compile_queue, _compilers[0], THREAD);
       if (ct == NULL) break;
+      ThreadsListHandle tlh;
+      if (!tlh.includes(ct)) break;
```
[the remark this answers is not in the document]
Comment (dcubed-ojdk) [reply src/hotspot/share/compiler/compileBroker.cpp:1136]: Okay. I've switched this one back to an assertion after the TLH.

File: src/hotspot/share/runtime/thread.cpp (hunk 15 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.cpp:2955]: I'm backing out both:
```
Threads::java_threads_do(JavaThreadIteratorWithHandle* jtiwh_p, ThreadClosure* tc)
Threads::threads_do(JavaThreadIteratorWithHandle* jtiwh_p, ThreadClosure* tc)
```

and I'm backing out the changes to src/hotspot/share/services/management.cpp.
By removing the Threads_lock usage, I'm allowing those closures to run in parallel
with other closures that are using the original Threads::java_threads_do() and
Threads::threads_do() calls that do use the Threads_lock. Those other uses might
be assuming exclusivity for their algorithms and my M&M changes break that.

File: src/hotspot/share/services/management.cpp (hunk 14 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/services/management.cpp:849]: I'm backing out the changes to src/hotspot/share/services/management.cpp.
By removing the Threads_lock usage, I'm allowing those closures to run in parallel
with other closures that are using the original Threads::java_threads_do() and
Threads::threads_do() calls that do use the Threads_lock. Those other uses might
be assuming exclusivity for their algorithms and my M&M changes break that.

File: src/hotspot/share/services/management.cpp (hunk 16 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/services/management.cpp:1704]: I'm backing out the changes to src/hotspot/share/services/management.cpp.

File: src/hotspot/share/runtime/thread.cpp (hunk 13 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.cpp:2568]: It has occurred to me that I should have moved all of the
protection checking logic and assertion checks into the
new Thread::is_JavaThread_protected() function. That will
allow JavaThread::get_thread_name() to become very simple.

**Review by dcubed-ojdk — COMMENTED** [review]
@dholmes-ora - Thanks for reviewing CR3. I've made more changes
and once my testing looks good, I'll checkpoint and push to update the PR.

**Comment by dcubed-ojdk** [comment]
You'll see that in the next update. Pushing it now...

File: src/hotspot/share/runtime/thread.cpp [hunk 19]
```diff
@@ -482,25 +482,49 @@ void Thread::check_for_dangling_thread_pointer(Thread *thread) {
...
 
-  // The target JavaThread is not on this ThreadsList or any of its
-  // nested ThreadsLists (if any) so it is not safe to query:
+  // Use this debug code with -XX:+UseNewCode to diagnose locations that
+  // are missing a ThreadsListHandle or other protection mechanism:
+  // guarantee(!UseNewCode, "current_thread=" INTPTR_FORMAT " is not protecting p="
+  //           INTPTR_FORMAT, p2i(current_thread), p2i(p));
+  guarantee(false, "current_thread=" INTPTR_FORMAT " is not protecting p="
+            INTPTR_FORMAT, p2i(current_thread), p2i(p));
+
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/thread.cpp:520]: This will be deleted before pushing - right?

File: src/hotspot/share/compiler/compileBroker.cpp [hunk 20]
```diff
@@ -1023,10 +1024,11 @@ void CompileBroker::init_compiler_sweeper_threads() {
     if (!UseDynamicNumberOfCompilerThreads || i == 0) {
       JavaThread *ct = make_thread(compiler_t, thread_handle, _c1_compile_queue, _compilers[0], THREAD);
       assert(ct != NULL, "should have been handled for initial thread");
+      ThreadsListHandle tlh;  // get_thread_name() depends on the TLH.
+      assert(tlh.includes(ct), "ct=" INTPTR_FORMAT " exited unexpectedly.", p2i(ct));
```
Comment (dholmes-ora) [inline src/hotspot/share/compiler/compileBroker.cpp:1028]: This can be placed inside the if-clause where the MutexLocker was.

File: src/hotspot/share/compiler/compileBroker.cpp [hunk 21]
```diff
@@ -1109,10 +1111,11 @@ void CompileBroker::possibly_add_compiler_threads(Thread* THREAD) {
 #endif
       JavaThread *ct = make_thread(compiler_t, compiler2_object(i), _c2_compile_queue, _compilers[1], THREAD);
       if (ct == NULL) break;
+      ThreadsListHandle tlh;  // get_thread_name() depends on the TLH.
+      assert(tlh.includes(ct), "ct=" INTPTR_FORMAT " exited unexpectedly.", p2i(ct));
```
Comment (dholmes-ora) [inline src/hotspot/share/compiler/compileBroker.cpp:1115]: This can be placed inside the if-clause where the MutexLocker was.

File: src/hotspot/share/compiler/compileBroker.cpp [hunk 22]
```diff
@@ -1129,10 +1132,11 @@ void CompileBroker::possibly_add_compiler_threads(Thread* THREAD) {
     for (int i = old_c1_count; i < new_c1_count; i++) {
       JavaThread *ct = make_thread(compiler_t, compiler1_object(i), _c1_compile_queue, _compilers[0], THREAD);
       if (ct == NULL) break;
+      ThreadsListHandle tlh;  // get_thread_name() depends on the TLH.
+      assert(tlh.includes(ct), "ct=" INTPTR_FORMAT " exited unexpectedly.", p2i(ct));
```
Comment (dholmes-ora) [inline src/hotspot/share/compiler/compileBroker.cpp:1136]: This can be placed inside the if-clause where the MutexLocker was.

**Review by dholmes-ora — APPROVED** [review]
Hi Dan,

I'm good with holding the Threads_lock remaining as a safe condition for use of get_thread_name().

Minor comments on code placement that should have been made earlier - sorry.

Thanks,
David

File: src/hotspot/share/runtime/thread.cpp (hunk 19 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/runtime/thread.cpp:520]: Yup. It's just for the current Tier[1-8] test cycle.

File: src/hotspot/share/compiler/compileBroker.cpp (hunk 20 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/compiler/compileBroker.cpp:1028]: True. That would be better.

File: src/hotspot/share/compiler/compileBroker.cpp (hunk 21 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/compiler/compileBroker.cpp:1115]: True. That would be better.

File: src/hotspot/share/compiler/compileBroker.cpp (hunk 22 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/compiler/compileBroker.cpp:1136]: True. That would be better.

**Review by dcubed-ojdk — COMMENTED** [review]
@dholmes-ora - Thanks for reviewing CR4.

**Review by dholmes-ora — APPROVED** [review]
Looks good!

Thanks,
David

**Comment by dcubed-ojdk** [comment]
@dholmes-ora - Thanks for yet another re-review.
@coleenp, @robehn and/or @sspitsyn - If any of you have the time, I could
use one more re-review.

**Review by robehn — APPROVED** [review]

**Review by sspitsyn — COMMENTED** [review]
Hi Dan,
The update looks okay, but I have one question.
So, the fix in jvmtiTrace.cpp is not needed anymore?
Thanks,
Serguei

**Comment by dcubed-ojdk** [comment]
@robehn - Thanks for the re-review!
@sspitsyn - Thanks for the re-review!
We backed out the jvmtiTrace.cpp changes because the semantics were
changed subtlety by switching to JavaThread::get_thread_name_string().
@dholmes-ora pointed out the change in behavior in one of his reviews.
